src/main/java/com/fasterxml/jackson/databind/util/clhm/ConcurrentLinkedHashMap.java

```diff
@@ -0,0 +1,1597 @@
+    void applyRead(Node<K, V> node) {
+        // An entry may be scheduled for reordering despite having been removed.
+        // This can occur when the entry was concurrently read while a writer was
+        // removing it. If the entry is no longer linked then it does not need to
+        // be processed.
+        if (evictionDeque.contains(node)) {
+            evictionDeque.moveToBack(node);
+        }
+    }
+
+    /** Drains the read buffer up to an amortized threshold. */
+    //@GuardedBy("evictionLock")
+    void drainWriteBuffer() {
+        for (int i = 0; i < WRITE_BUFFER_DRAIN_THRESHOLD; i++) {
+            final Runnable task = writeBuffer.poll();
+            if (task == null) {
+                break;
+            }
+            task.run();
+        }
+    }
+
+    /**
+     * Attempts to transition the node from the <tt>alive</tt> state to the
+     * <tt>retired</tt> state.
+     *
+     * @param node the entry in the page replacement policy
+     * @param expect the expected weighted value
+     * @return if successful
+     */
+    boolean tryToRetire(Node<K, V> node, WeightedValue<V> expect) {
+        if (expect.isAlive()) {
+            final WeightedValue<V> retired = new WeightedValue<V>(expect.value, -expect.weight);
+            return node.compareAndSet(expect, retired);
+        }
+        return false;
+    }
+
+    /**
+     * Atomically transitions the node from the <tt>alive</tt> state to the
+     * <tt>retired</tt> state, if a valid transition.
+     *
+     * @param node the entry in the page replacement policy
+     */
+    void makeRetired(Node<K, V> node) {
+        for (;;) {
+            final WeightedValue<V> current = node.get();
+            if (!current.isAlive()) {
+                return;
+            }
+            final WeightedValue<V> retired = new WeightedValue<V>(current.value, -current.weight);
+            if (node.compareAndSet(current, retired)) {
+                return;
+            }
+        }
+    }
+
+    /**
+     * Atomically transitions the node to the <tt>dead</tt> state and decrements
+     * the <tt>weightedSize</tt>.
+     *
+     * @param node the entry in the page replacement policy
+     */
+    //@GuardedBy("evictionLock")
+    void makeDead(Node<K, V> node) {
+        for (;;) {
+            WeightedValue<V> current = node.get();
+            WeightedValue<V> dead = new WeightedValue<V>(current.value, 0);
+            if (node.compareAndSet(current, dead)) {
+                weightedSize.lazySet(weightedSize.get() - Math.abs(current.weight));
+                return;
+            }
+        }
+    }
+
+    /** Notifies the listener of entries that were evicted. */
+    void notifyListener() {
+        Node<K, V> node;
+        while ((node = pendingNotifications.poll()) != null) {
+            listener.onEviction(node.key, node.getValue());
+        }
+    }
+
+    /** Adds the node to the page replacement policy. */
+    final class AddTask implements Runnable {
```

Comment:
Can perhaps get rid of this too (unless it is used even when not run from separate thread)

Reply:
These tasks are the buffer replay actions, not for an Executor.

Reply:
Ok.